_posts/2020-11-12-deconvolution.md

```diff
@@ -0,0 +1,436 @@
+---
+layout: post
+title: Image Analysis Redux
+tagline: Dask + CuPy + RL
+author: John Kirkham (NVIDIA) and Ben Zaitlen (NVIDIA)
+theme: twitter
+---
+{% include JB/setup %}
+
+Summary
+-------
+
+[Last year](https://blog.dask.org/2019/08/09/image-itk) we experimented with
+Dask/ITK/Scikit-Image to perform large scale image analysis on a stack of 3D
+images.  Specifically, we looked at deconvolution, a common method to _deblur_
+images.  Now, a year later, we return to these experiments with a better
+understanding of how Dask and CuPy can interact, enhanced serialization
+methods, and support from the open-source community. This post looks at the
+following:
+
+1. Implementing a common deconvolution method for CPU + GPU
+2. Leveraging Dask to perform deconvolution on a larger dataset
+3. Exploring the results with the Napari image viewer
+
+Image Analysis Redux
+--------------------
+
+Previously we used the [Richardson Lucy
+(RL)](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution)
+deconvolution algorithm from ITK and
+[Scikit-Image](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/deconvolution.py#L329).
+We left off at theorizing how GPUs could potentially help accelerate these
+workflows.  Starting with Scikit-Image's implementation, we naively tried
+replacing `scipy.signal.convolve` calls with `cupyx.scipy.ndimage.convolve`,
+and while performance improved, it did not improve _significantly_ -- that is,
+we did not get the 100X speed we were looking for.
+
+As it often turns out in mathematics a problem that can be inefficient to solve
+in one representation can often be made more efficent by transforming the data
+beforehand. In this new representation we can solve the same problem
+(convolution in this case) more easily before transforming the result back into
+a more familiar representation. When it comes to convolution, the
+transformation we apply is called [Fast-Fourier Transform
+(FFT)](https://en.wikipedia.org/wiki/Fast_Fourier_transform). Once this is
+applied we are able to convolve data using a simple multiplication.
```

Comment:
Apologies if this comes across as pedantic, but I just re-read this after looking at a few hours: there are still some phrases that I consider slightly unfortunate. You mention at several points that one can speed up convolutions by going to the Fourier domain (I just picked one example here, that statement is present multiple times with slight variation). 
Convolutions are a mathematical operation typically reprenset by nested integrals (or sums in the discrete case). We are not really speeding up convolutions by going to the Fourier domain, we are avoiding them altogether. The result (when transformed back into the spatial domain) is (boundary effects and other numerical issues aside) the same.
So what is sped up are signal filtering operations, through *not* performing convolutions.
The algorithm in its totality should still be called "deconvolution" though as it is not referring to the specific mathematical operation, even if implemented without explicit convolutions.

Reply:
However, I might be too sensitive here, the wikipedia article https://en.wikipedia.org/wiki/Convolution_theorem uses similar phrasing to yours: 
```
This formulation is especially useful for implementing a numerical convolution on a computer: The standard convolution algorithm has quadratic computational complexity. With the help of the convolution theorem and the fast Fourier transform, the complexity of the convolution can be reduced from O ( n 2 ) {\displaystyle O\left(n^{2}\right)} {\displaystyle O\left(n^{2}\right)} to O ( n log ⁡ n ) {\displaystyle O\left(n\log n\right)} {\displaystyle O\left(n\log n\right)},
```
So maybe disregard my overly pedantic interpreation above.

Reply:
I don't know that I would agree we are avoiding convolutions. Multiplying a kernel by an image in Fourier space is the same as performing the nested integral in real space. Both constitute a convolution. How we turn the crank to get that result may differ, but the result is still a convolution of the two.

Reply:
Ok. I stand corrected.
Quoting wikipedia again:
*The term convolution refers to both the result function and to the process of computing it.*
I was too rigorous above in referring only the process as convolution.  
All good.

Reply:
No worries. We appreciate the thoroughness 🙂